README.md
# live-coding

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-01-28
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-01-28
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/live-coding/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/live-coding/compare/0.1.0...0.1.1


src/live_coding/core.clj
(ns live-coding.core
  (:use
   [clojure.pprint]
   [overtone.live])
  (:require [shadertone.tone :as s]
))

(load-file "src/live_coding/drums.clj")
(load-file "src/live_coding/synths.clj")
(load-file "src/live_coding/channels.clj")

;; start shader
(s/start "src/live_coding/shaders/unknown_pleasures.glsl"
         :width 2600 :height 1400)

;; metronome
(def metro (metronome 120))

;; sec 16 count_by 0.25 = 64
(defn player [beat]
  (let [sec 16]
    ;; (ch1 beat sec)
    ;; (ch2 beat sec)
    ;; (ch3 beat sec)
    ;; (ch4 beat sec)
    ;; (ch5 beat sec)
    ;; (ch6 beat sec)
    ;; (ch7 beat sec)

    (ch10 beat sec)
    (ch11 beat sec)
    (ch13 beat sec)
    (apply-by (metro (+ sec beat)) #'player (+ sec beat) [])
))

(player (metro))
(stop)

(bubbles :base-freq 520)
(kill bubbles)


(defn ch10 [beat sec]
  (let [beats (range 0 sec 0.25)
        pattern [1 0 0 0 1 0 0 0 1 0 0 0 1 0 1 1]
        pattern (flatten (repeat (/ sec 4) pattern))
        pairs (map vector beats pattern)]

    (doseq [pair pairs]
      (let [b (first pair) n (last pair)]
        (at (metro (+ b beat)) (kick2 :amp n))
    ))
))

(defn ch11 [beat sec]
  (let [beats (range 0 sec 0.25)
        pattern [0 1 0 1 1 1 0 0 0 0 0 0 0 1 0 1]
        pattern (flatten (repeat (/ sec 4) pattern))
        pairs (map vector beats pattern)]

    (doseq [pair pairs]
      (let [b (first pair) n (last pair)]
        (at (metro (+ b beat)) (closed-hat :amp (* n 0.2)))
    ))
))

(defn ch13 [beat sec]
  (let [amp 1.5
        release 3]

    (at (metro (+ 0.0 beat))
        (mix [(simple-flute :freq (midi->hz 58) :amp amp :release release)
              (simple-flute :freq (midi->hz 60) :amp amp :release release)
              (simple-flute :freq (midi->hz 67) :amp amp :release release)]))

    (at (metro (+ 4.0 beat))
        (mix [(simple-flute :freq (midi->hz 56) :amp amp :release release)
              (simple-flute :freq (midi->hz 58) :amp amp :release release)
              (simple-flute :freq (midi->hz 65) :amp amp :release release)]))

    (at (metro (+ 8.0 beat))
        (mix [(simple-flute :freq (midi->hz 56) :amp amp :release release)
              (simple-flute :freq (midi->hz 58) :amp amp :release release)
              (simple-flute :freq (midi->hz 65) :amp amp :release release)]))

    (at (metro (+ 12.0 beat))
        (mix [(simple-flute :freq (midi->hz 55) :amp amp :release release)
              (simple-flute :freq (midi->hz 57) :amp amp :release release)
              (simple-flute :freq (midi->hz 62) :amp amp :release release)]))
))




src/live_coding/drums.clj
;; kicks
(definst kick1
  [freq       {:default 80 :min 50 :max 140 :step 1}
   env-ratio  {:default 3 :min 1.2 :max 8.0 :step 0.1}
   freq-decay {:default 0.02 :min 0.001 :max 1.0 :step 0.001}
   amp-decay  {:default 0.8 :min 0.001 :max 1.2 :step 0.001}]
  (let [fenv (* (env-gen (envelope [env-ratio 1] [freq-decay] :exp)) freq)
        aenv (env-gen (perc 0.005 amp-decay) :action FREE)]
    (* (sin-osc fenv (* 0.5 Math/PI)) aenv)))

(definst kick2 [freq      {:default 80 :min 10 :max 20000 :step 1}
                amp       {:default 0.8 :min 0.001 :max 1.0 :step 0.001}
                mod-freq  {:default 5 :min 0.001 :max 10.0 :step 0.01}
                mod-index {:default 5 :min 0.001 :max 10.0 :step 0.01}
                sustain   {:default 0.4 :min 0.001 :max 1.0 :step 0.001}
                noise     {:default 0.025 :min 0.001 :max 1.0 :step 0.001}]
  (let [pitch-contour (line:kr (* 2 freq) freq 0.02)
        drum (lpf (sin-osc pitch-contour (sin-osc mod-freq (/ mod-index 1.3))) 1000)
        drum-env (env-gen (perc 0.005 sustain) :action FREE)
        hit (hpf (* noise (white-noise)) 500)
        hit (lpf hit (line 6000 500 0.03))
        hit-env (env-gen (perc))]
    (* amp (+ (* drum drum-env) (* hit hit-env)))))

(definst kick3
  [freq {:default 80 :min 40 :max 140 :step 1}
   amp {:default 0.3 :min 0.001 :max 1 :step 0.001}]
  (let [sub-osc   (sin-osc freq)
        sub-env   (line 1 0 0.7 FREE)
        click-osc  (lpf (white-noise) 1500)
        click-env  (line 1 0 0.02)
        sub-out   (* sub-osc sub-env)
        click-out (* click-osc click-env)]
    (* amp (+ sub-out click-out))))

(definst kick4
  [freq   {:default 80 :min 40 :max 140 :step 1}
   amp    {:default 0.3 :min 0.001 :max 1 :step 0.001}
   attack {:default 0.001 :min 0.001 :max 1.0 :step 0.001}
   decay  {:default 0.4 :min 0.001 :max 1 :step 0.001}]
  (let [env (env-gen (perc attack decay) :action FREE)
        snd (sin-osc freq (* Math/PI 0.5))
        snd (* amp env snd)]
    snd))

;; hats
(definst open-hat
  [amp    {:default 0.3 :min 0.001 :max 1 :step 0.01}
   t      {:default 0.3 :min 0.1 :max 1.0 :step 0.01}
   low    {:default 6000 :min 3000 :max 12000 :step 1}
   hi     {:default 2000 :min 1000 :max 8000 :step 1}]
  (let [low (lpf (white-noise) low)
        hi (hpf low hi)
        env (line 1 0 t :action FREE)]
    (* amp env hi)))

(definst closed-hat
  [amp    {:default 0.3 :min 0.001 :max 1 :step 0.01}
   t      {:default 0.1 :min 0.1 :max 1.0 :step 0.01}
   low    {:default 6000 :min 3000 :max 12000 :step 1}
   hi     {:default 2000 :min 1000 :max 8000 :step 1}]
  (let [low (lpf (white-noise) low)
        hi (hpf low hi)
        env (line 1 0 t :action FREE)]
    (* amp env hi)))

(definst closed-hat2
  [amp    {:default 0.3 :min 0.001 :max 1.0 :step 0.001}
   attack {:default 0.001 :min 0.001 :max 1.0 :step 0.0001}
   decay  {:default 0.07 :min 0.001 :max 1.0 :step 0.001}]
  (let [env (env-gen (perc attack decay) 1 1 0 1 FREE)
        noise (white-noise)
        sqr (* (env-gen (perc 0.01 0.04)) (pulse 880 0.2))
        filt (bpf (+ sqr noise) 9000 0.5)]
    (* 0.5 amp env filt)))

;; snares
(definst snare
  [freq   {:default 405 :min 100 :max 1000 :step 1}
   amp    {:default 0.3 :min 0.001 :max 1 :step 0.01}
   sustain {:default 0.1 :min 0.01 :max 1.0 :step 0.001}
   decay  {:default 0.1 :min 0.1 :max 1.0 :step 0.01}
   drum-amp 0.25
   crackle-amp 40
   tightness 1000]
  (let [drum-env  (env-gen (perc 0.005 sustain) :action FREE)
        drum-osc  (mix (* drum-env (sin-osc [freq (* freq 0.53)])))
        drum-s3   (* drum-env (pm-osc (saw (* freq 0.85)) 184 (/ 0.5 1.3)))
        drum      (* drum-amp (+ drum-s3 drum-osc))
        noise     (* 0.1 (lf-noise0 20000))
        noise-env (env-gen (perc 0.005 sustain) :action FREE)
        filtered  (* 0.5 (brf noise 8000 0.1))
        filtered  (* 0.5 (brf filtered 5000 0.1))
        filtered  (* 0.5 (brf filtered 3600 0.1))
        filtered  (* (brf filtered 2000 0.0001) noise-env)
        resonance (* (resonz filtered tightness) crackle-amp)]
    (* amp (+ drum resonance))))

(definst noise-snare
  [freq   {:default 1000 :min 100 :max 10000 :step 1}
   amp    {:default 0.3 :min 0.001 :max 1 :step 0.01}
   decay  {:default 0.1 :min 0.1 :max 1.0 :step 0.01}]
  (let [env (env-gen (perc 0 decay) :action FREE)
        snd (bpf (gray-noise) freq 3)]
    (* snd env amp)))

;; toms
(definst tom
  [freq {:default 90 :min 50 :max 400 :step 1}
   amp {:default 0.3 :min 0.001 :max 1 :step 0.01}
   sustain {:default 0.4 :min 0.01 :max 1.0 :step 0.001}
   mode-level {:default 0.25 :min 0.01 :max 1.0 :step 0.001}
   timbre {:default 1 :min 0.1 :max 5.0 :step 0.1}
   stick-level {:default 0.2 :min 0.0 :max 1.0 :step 0.1}]
  (let [env (env-gen (perc 0.005 sustain) :action FREE)
        s1 (* 0.5 env (sin-osc (* freq 0.8)))
        s2 (* 0.5 env (sin-osc freq))
        s3 (* 5 env (sin-osc (saw (* 0.9 freq))
                             (* (sin-osc (* freq 0.85))
                                (/ timbre 1.3))))
        mix (* mode-level (+ s1 s2 s3))
        stick (* stick-level
                 (env-gen (perc 0.001 0.01))
                 (crackle 2.0))
        mix2 (* amp (+ mix stick))]
    mix2))

;; percussive
(definst clap
  [low {:default 7500 :min 100 :max 10000 :step 1}
   hi  {:default 1500 :min 100 :max 10000 :step 1}
   amp {:default 0.3 :min 0.001 :max 1 :step 0.01}
   decay {:default 0.6 :min 0.1 :max 0.8 :step 0.001}]
  (let [noise      (bpf (lpf (white-noise) low) hi)
        clap-env   (line 1 0 decay :action FREE)
        noise-envs (map #(envelope [0 0 1 0] [(* % 0.01) 0 0.04]) (range 8))
        claps      (apply + (* noise (map env-gen noise-envs)))]
    (* amp (* claps clap-env))))

(definst haziti-clap
  [freq   {:default 44.77 :min 20 :max 400 :step 1}
   attack {:default 0.036 :min 0.00001 :max 2 :step 0.0001}
   decay  {:default 1.884 :min 0.00001 :max 2 :step 0.0001}
   rq     {:default 0.08 :min 0.01 :max 1 :step 0.01}
   amp    {:default 0.8 :min 0.01 :max 1 :step 0.01}]
  (let [noiz (white-noise)
        bfreq (* freq (abs (lf-noise0 80)))
        filt (* 4 (bpf (rhpf noiz 4064.78 rq) bfreq (* 1 rq)))
        env  (x-line 1 0.001 decay :action FREE)
        wave (lf-tri (* (abs (lf-noise0:kr 699)) 4400))
        wenv (env-gen (perc 0.00001 0.008))
        skip (* wave wenv)]
    (* amp (+ (* env filt) skip))))



src/live_coding/channels.clj
(defn ch1 [beat sec]
  (let [beats (range 0 sec 0.25)
        pattern [1 0 1 1 0 0 1 0 0 0 0 0 0 0 0 0]
        pattern (flatten (repeat (/ sec 4) pattern))
        pairs (map vector beats pattern)]

    (doseq [pair pairs]
      (let [b (first pair) n (last pair)]
        (at (metro (+ b beat)) (kick2 :amp n))
    ))
))

(defn ch2 [beat sec]
  (let [beats (range 0 sec 0.25)
        pattern [0 0 0 0 0 0 0 0 1 1 1 0 1 0 1 0]
        pattern (flatten (repeat (/ sec 4) pattern))
        pairs (map vector beats pattern)]

    (doseq [pair pairs]
      (let [b (first pair) n (last pair)]
        (at (metro (+ b beat)) (clap :amp (* n 0.1)))
    ))
))

(defn ch3 [beat sec]
  (let [beats   (range 0 sec 0.25)
        pattern [63 0 73 70 0 0 66 0 0 0 75 73 0 0 78 0]
        pattern (flatten (repeat (/ sec 4) pattern))
        pairs    (map vector beats pattern)]

    (doseq [pair pairs]
      (let [b (first pair) n (last pair)]
        (at (metro (+ b beat)) (vintage-bass :note n :amp 0.6))
    ))
))

(defn ch4 [beat sec]
  (let [n (/ sec 2)]
    (at (metro (+ 0 beat)) (rise-fall-pad :freq (midi->hz 75)))
    (at (metro (+ n beat)) (rise-fall-pad :freq (midi->hz 73)))
    )
)

(defn ch5 [beat sec]
  (let [n (/ sec 2)]
    (at (metro (+ 1 beat)) (rise-fall-pad :freq (midi->hz 65)))
    (at (metro (+ n 1 beat)) (rise-fall-pad :freq (midi->hz 61)))
    )
)

(defn ch6 [beat sec]
  (let [n (/ sec 2)]
    (at (metro (+ 2 beat)) (rise-fall-pad :freq (midi->hz 87)))
    (at (metro (+ n 2 beat)) (rise-fall-pad :freq (midi->hz 85)))
    )
)

(defn ch7 [beat sec]
  (let [n (/ sec 2)]
    (at (metro (+ 1 beat)) (whoahaha :freq (midi->hz 30) :amp 0.2))
    (at (metro (+ n 1 beat)) (whoahaha :freq (midi->hz 30) :amp 0.2))
    )
)

(defn ch8 [beat sec]
  (let [beats   (range 0 sec 0.25)
        pattern [63 0 73 70 0 0 66 0 0 0 75 73 0 0 78 0]
        pattern (flatten (repeat (/ sec 4) pattern))
        pairs    (map vector beats pattern)]

    (doseq [pair pairs]
      (let [b (first pair) n (last pair)]
        (at (metro (+ b beat)) (vintage-bass :note n :amp 0.6))
    ))

))

(defn ch10 [beat sec]
  (let [beats (range 0 sec 0.25)
        pattern [1 0 1 0 0 0 0 0 1 0 0 0 0 0 1 1]
        pattern (flatten (repeat (/ sec 4) pattern))
        pairs (map vector beats pattern)]

    (doseq [pair pairs]
      (let [b (first pair) n (last pair)]
        (at (metro (+ b beat)) (kick2 :amp n))
    ))
))

(defn ch11 [beat sec]
  (let [beats (range 0 sec 0.25)
        pattern [0 0 0 0 1 0 0 0 0 0 0 0 1 0 1 0]
        pattern (flatten (repeat (/ sec 4) pattern))
        pairs (map vector beats pattern)]

    (doseq [pair pairs]
      (let [b (first pair) n (last pair)]
        (at (metro (+ b beat)) (open-hat :amp (* n 0.2)))
    ))
))



src/live_coding/synths.clj
(definst simple-flute [freq 880
                       amp 0.5
                       attack 0.4
                       decay 0.5
                       sustain 0.8
                       release 1
                       gate 1
                       out 0]
  (let [env  (env-gen (adsr attack decay sustain release) gate :action FREE)
        mod1 (lin-lin:kr (sin-osc:kr 6) -1 1 (* freq 0.99) (* freq 1.01))
        mod2 (lin-lin:kr (lf-noise2:kr 1) -1 1 0.2 1)
        mod3 (lin-lin:kr (sin-osc:kr (ranged-rand 4 6)) -1 1 0.5 1)
        sig (distort (* env (sin-osc [freq mod1])))
        sig (* amp sig mod2 mod3)
        env2 (env-gen (perc attack release))
        sig (* env2 sig)]
    sig))

(definst supersaw [freq 440 amp 1 attack 0.1 release 0.3]
  (let [input  (lf-saw freq)
        shift1 (lf-saw 4)
        shift2 (lf-saw 7)
        shift3 (lf-saw 5)
        shift4 (lf-saw 2)
        comp1  (> input shift1)
        comp2  (> input shift2)
        comp3  (> input shift3)
        comp4  (> input shift4)
        output (+ (- input comp1) (- input comp2) (- input comp3) (- input comp4))
        output (- output input)
        output (leak-dc:ar (* output 0.25))
        env (env-gen (perc 0.1 0.3))
        output (* env output)]
    (* amp output)))

(definst ping
  [note   {:default 72   :min 0     :max 120 :step 1}
   attack {:default 0.02 :min 0.001 :max 1   :step 0.001}
   decay  {:default 0.3  :min 0.001 :max 1   :step 0.001}]
  (let [snd (sin-osc (midicps note))
        env (env-gen (perc attack decay) :action FREE)]
    (* 0.8 env snd)))

(definst tb303
  [note       {:default 60 :min 0 :max 120 :step 1}
   wave       {:default 1 :min 0 :max 2 :step 1}
   r          {:default 0.8 :min 0.01 :max 0.99 :step 0.01}
   attack     {:default 0.01 :min 0.001 :max 4 :step 0.001}
   decay      {:default 0.1 :min 0.001 :max 4 :step 0.001}
   sustain    {:default 0.6 :min 0.001 :max 0.99 :step 0.001}
   release    {:default 0.01 :min 0.001 :max 4 :step 0.001}
   cutoff     {:default 100 :min 1 :max 20000 :step 1}
   env-amount {:default 0.01 :min 0.001 :max 4 :step 0.001}
   amp        {:default 2 :min 0 :max 5 :step 0.01}]
  (let [freq       (midicps note)
        freqs      [freq (* 1.01 freq)]
        vol-env    (env-gen (adsr attack decay sustain release)
                            (line:kr 1 0 (+ attack decay release))
                            :action FREE)
        fil-env    (env-gen (perc))
        fil-cutoff (+ cutoff (* env-amount fil-env))
        waves      (* vol-env
                      [(saw freqs)
                       (pulse freqs 0.5)
                       (lf-tri freqs)])
        selector   (select wave waves)
        filt       (rlpf selector fil-cutoff r)]
    (* amp filt)))

(definst rise-fall-pad
  [freq 440 t 4 amt 0.3 amp 0.8]
  (let [f-env      (env-gen (perc t t) 1 1 0 1 FREE)
        src        (saw [freq (* freq 1.01)])
        signal     (rlpf (* 0.3 src)
                         (+ (* 0.6 freq) (* f-env 2 freq)) 0.2)
        k          (/ (* 2 amt) (- 1 amt))
        distort    (/ (* (+ 1 k) signal) (+ 1 (* k (abs signal))))
        gate       (pulse (* 2 (+ 1 (sin-osc:kr 0.05))))
        compressor (compander distort gate 0.01 1 0.5 0.01 0.01)
        dampener   (+ 1 (* 0.5 (sin-osc:kr 0.5)))
        reverb     (free-verb compressor 0.5 0.5 dampener)
        echo       (comb-n reverb 0.4 0.3 0.5)]
    (* amp echo)))

(definst overpad
  [note 60 amp 0.7 attack 0.001 release 2]
  (let [freq  (midicps note)
        env   (env-gen (perc attack release) :action FREE)
        f-env (+ freq (* 3 freq (env-gen (perc 0.012 (- release 0.1)))))
        bfreq (/ freq 2)
        sig   (apply +
                     (concat (* 0.7 (sin-osc [bfreq (* 0.99 bfreq)]))
                             (lpf (saw [freq (* freq 1.01)]) f-env)))
        audio (* amp env sig)]
    audio))

(definst bass
  [freq 120 t 0.6 amp 0.8]
  (let [env  (env-gen (perc 0.08 t) :action FREE)
        src  (saw [freq (* 0.98 freq) (* 2.015 freq)])
        src  (clip2 (* 1.3 src) 0.8)
        sub  (sin-osc (/ freq 2))
        filt (resonz (rlpf src (* 4.4 freq) 0.09) (* 2.0 freq) 2.9)]
    (* env amp (fold:ar (distort (* 1.3 (+ filt sub))) 0.08))))

(definst vintage-bass
  [note 40 velocity 80 t 0.6 amp 1 gate 1 release 1]
  (let [freq     (midicps note)
        sub-freq (midicps (- note 12))
        velocity (/ velocity 127.0)
        sawz1    (* 0.275 (saw [freq (* 1.01 freq)]))
        sawz2    (* 0.75 (saw [(- freq 2) (+ 1 freq)]))
        sqz      (* 0.3 (pulse [sub-freq (- sub-freq 1)]))
        mixed    (* 5 (+ sawz1 sawz2 sqz))
        env      (env-gen (adsr 0.1 3.3 0.4 0.8) gate :action FREE)
        filt     (* env (moog-ff mixed (* velocity env (+ freq 200)) 2.2))
        open-env (env-gen (perc 0.1 release) :action FREE)
        filt     (* open-env filt)]
    (* amp filt)))


; B3 modeled a church organ using additive synthesis of 9 sin oscillators
; * Octave under root
; *	Fifth over root
; * Root
; * Octave over root
; * Octave and a fifth over root
; * Two octaves over root
; * Two octaves and a major third over root
; * Two octaves and a fifth over root
; * Three octaves over root
; Work in progress...  just getting started
(comment definst b3
  [note 60 a 0.01 d 3 s 1 r 0.01]
  (let [freq  (midicps note)
        waves (sin-osc [(* 0.5 freq)
                        freq
                        (* (/ 3 2) freq)
                        (* 2 freq)
                        (* freq 2 (/ 3 2))
                        (* freq 2 2)
                        (* freq 2 2 (/ 5 4))
                        (* freq 2 2 (/ 3 2))
                        (* freq 2 2 2)])
        snd   (apply + waves)
        env   (env-gen (adsr a d s r) :action FREE)]
    (* env snd 0.1)))

; Experimenting with Karplus Strong synthesis...
(definst ks1
  [note  {:default 60  :min 10   :max 120  :step 1}
   amp   {:default 0.8 :min 0.01 :max 0.99 :step 0.01}
   dur   {:default 2   :min 0.1  :max 4    :step 0.1}
   decay {:default 30  :min 1    :max 50   :step 1}
   coef  {:default 0.3 :min 0.01 :max 2    :step 0.01}]
  (let [freq (midicps note)
        noize (* 0.8 (white-noise))
        dly (/ 1.0 freq)
        plk   (pluck noize 1 (/ 1.0 freq) dly
                     decay
                     coef)
        dist (distort plk)
        filt (rlpf dist (* 12 freq) 0.6)
        clp (clip2 filt 0.8)
        reverb (free-verb clp 0.4 0.8 0.2)]
    (* amp (env-gen (perc 0.0001 dur) :action FREE) reverb)))

(definst ks1-demo
  [note 60 amp 0.8 gate 1]
  (let [freq (midicps note)
        noize (* 0.8 (white-noise))
        dly (/ 1.0 freq)
        plk   (pluck noize gate (/ 1.0 freq) dly
                     (mouse-x 0.1 50)
                     (mouse-y 0.0001 0.9999))
        dist (distort plk)
        filt (rlpf dist (* 12 freq) 0.6)
        reverb (free-verb filt 0.4 0.8 0.2)]
    (* amp (env-gen (perc 0.0001 2) :action FREE) reverb)))

(definst ks-stringer
  [freq 440 rate 6]
  (let [noize (* 0.8 (white-noise))
        trig  (dust rate)
        coef  (mouse-x -0.999 0.999)
        delay (/ 1.0 (* (mouse-y 0.001 0.999) freq))
        plk   (pluck noize trig (/ 1.0 freq) delay 10 coef)
        filt (rlpf plk (* 12 freq) 0.6)]
    (* 0.8 filt)))

(definst fm-demo
  [note 60 amp 0.2 gate 0]
  (let [freq (midicps note)
        osc-a (* (sin-osc (mouse-x 20 3000))
                 0.3)
        osc-b (* amp (sin-osc (* (mouse-y 3000 0) osc-a)))]
    osc-a))

; From the SC2 examples included with SC
; Don't think it's quite there, but almost...
(definst harmonic-swimming
  [amp 0.5]
  (let [freq     100
        partials 20
        z-init   0
        offset   (line:kr 0 -0.02 60)
        snd (loop [z z-init
                   i 0]
              (if (= partials i)
                z
                (let [f (clip:kr (mul-add
                                   (lf-noise1:kr [(+ 6 (rand 4))
                                                  (+ 6 (rand 4))])
                                   0.2 offset))
                      src  (f-sin-osc (* freq (inc i)))
                      newz (mul-add src f z)]
                  (recur newz (inc i)))))]
    (out 10 (pan2 (* amp snd)))))

(definst whoahaha
  [freq 440 dur 5 osc 100 mul 1000 amp 1]
  (let [freqs [freq (* freq 1.0068) (* freq 1.0159)]
        sound (resonz (saw (map #(+ % (* (sin-osc osc) mul)) freqs))
                      (x-line 10000 10 25)
                      (line 1 0.05 25))
        sound (apply + sound)]
    (* amp
       (* (lf-saw:kr (line:kr 13 17 3)) (line:kr 1 0 dur FREE) sound))
))

(definst bubbles
  [bass-freq 80]
  (let [bub (+ bass-freq (* 3 (lf-saw:kr [8 7.23])))
        glis (+ bub (* 24 (lf-saw:kr 0.4 0)))
        freq (midicps glis)
        src (* 0.04 (sin-osc freq))
        zout (comb-n src :decay-time 4)]
    zout))

; // Originally from the STK instrument models...
(comment definst bowed
  [note 60 velocity 80 gate 1 amp 1
   bow-offset 0 bow-slope 0.5 bow-position 0.75 vib-freq 6.127 vib-gain 0.2]
  (let [freq         (midicps note)
        velocity     (/ velocity 127)
        beta-ratio   (+ 0.027236 (* 0.2 bow-position))
        base-delay   (reciprocal freq)
        [fb1 fb2]    (local-in 2)
        vibrato      (* (sin-osc vib-freq) vib-gain)
        neck-delay   (+ (* base-delay (- 1 beta-ratio)) (* base-delay vibrato))
        neck         (delay-l fb1 0.05 neck-delay)
        nut-refl     (neg neck)
        bridge       (delay-l fb2 0.025 (* base-delay beta-ratio))
        string-filt  (one-pole (* bridge 0.95) 0.55)
        bridge-refl  (neg string-filt)
        adsr         (* amp (env-gen (adsr 0.02 3.005 1.0 0.01) gate :action FREE))
        string-vel   (+ bridge-refl nut-refl)
        vel-diff     (- adsr string-vel)
        slope        (- 5.0 (* 4 bow-slope))
        bow-table    (clip:ar (pow (abs (+ (* (+ vel-diff bow-offset) slope) 0.75 )) -4) 0 1)
        new-vel       (* vel-diff bow-table)]
   (local-out (+ [bridge-refl nut-refl] new-vel))
   (resonz (* bridge 0.5) 500 0.85)))

(comment definst flute
  [gate 1 freq 440 amp 1.0 endreflection 0.5 jetreflection 0.5
   jetratio 0.32 noise-gain 0.15 vibfreq 5.925 vib-gain 0.0 amp 1.0]
  (let [nenv           (env-gen (linen 0.2 0.03 0.5 0.5) gate :action FREE)
        adsr           (+ (* amp 0.2) (env-gen (adsr 0.005 0.01 1.1 0.01) gate :action FREE))
        noise          (* (white-noise) noise-gain)
        vibrato        (sin-osc vibfreq 0 vib-gain)
        delay          (reciprocal (* freq 0.66666))
        lastout        (local-in 1)
        breathpressure (* adsr (+ noise, vibrato))
        filter         (leak-dc (one-pole (neg lastout) 0.7))
        pressurediff   (- breathpressure (* jetreflection filter))
        jetdelay       (delay-l pressurediff 0.025 (* delay jetratio))
        jet            (clip2 (* jetdelay (- (squared jetdelay) 1.0)) 1.0)
        boredelay      (delay-l (+ jet (* endreflection filter) 0.05 delay))]
    (local-out boredelay)
    (* 0.3 boredelay amp nenv)))

; SynthDef(\flute, { arg out=0, gate=1, freq=440, amp=1.0, endReflection=0.5, jetReflection=0.5, jetRatio=0.32, noiseGain=0.15, vibFreq=5.925, vibGain=0.0, outputGain=1.0;
;
;   var nenv = EnvGen.ar(Env.linen(0.2, 0.03, 0.5, 0.5), gate, doneAction: 2);
;   var adsr = (amp*0.2) + EnvGen.ar(Env.adsr(0.005, 0.01, 1.1, 0.01), gate, doneAction: 2);
;   var noise = WhiteNoise.ar(noiseGain);
;   var vibrato = SinOsc.ar(vibFreq, 0, vibGain);
;
;   var delay = (freq*0.66666).reciprocal;
;   var lastOut = LocalIn.ar(1);
;   var breathPressure = adsr*Mix([1.0, noise, vibrato]);
;   var filter = LeakDC.ar(OnePole.ar(lastOut.neg, 0.7));
;   var pressureDiff = breathPressure - (jetReflection*filter);
;   var jetDelay = DelayL.ar(pressureDiff, 0.025, delay*jetRatio);
;   var jet = (jetDelay * (jetDelay.squared - 1.0)).clip2(1.0);
;   var boreDelay = DelayL.ar(jet + (endReflection*filter), 0.05, delay);
;   LocalOut.ar(boreDelay);
;   Out.ar(out, 0.3*boreDelay*outputGain*nenv);
; }).store;
;
; s = Synth("flute", ["freq", 220]);
; s.set("gate", 0);
;
; SynthDef(\blowbotl, { arg out=0, amp=1.0, freq=440, rq=0.0, gate=1, noise=0.0, vibFreq=5.2, vibGain=0.0;
;   var lastOut = LocalIn.ar(1);
;   var adsr = amp*EnvGen.ar(Env.adsr(0.005, 0.01, 1.0, 0.010), gate, doneAction: 2);
;   var vibrato = SinOsc.ar(vibFreq, 0, vibGain);
;   var pressureDiff = (adsr+vibrato) - lastOut;
;   var jet = (pressureDiff * (pressureDiff.squared - 1.0)).clip2(1.0);
;   var randPressure = WhiteNoise.ar(noise)*adsr*(1.0 + pressureDiff);
;
;   var resonator = Resonz.ar(adsr+randPressure - (jet*pressureDiff), freq, rq);
;   LocalOut.ar(resonator);
;   Out.ar(out, LeakDC.ar(resonator));
; }).store
;
; f = Synth(\blowbotl);
; f.set("freq", 100);
; f.free;

; SynthDef(\voicform, { arg out=0, gate=1, freq=440, amp=0.3, voiceGain=1.0, noiseGain=0.0, sweepRate=0.001;
;
;   var voiced = Pulse.ar(freq, 0.1, voiceGain);
;   var onezero = OneZero.ar(voiced, -0.9);
;   var onepole = OnePole.ar(onezero, 0.97 - (amp*0.2));
;   var noise = WhiteNoise.ar(noiseGain*0.1);
;   var excitation = onepole + noise;
;
;   var ffreqs = Control.names([\ffreq]).kr([770, 1153, 2450, 3140]);
;   var fradii = Control.names([\bw]).kr([0.950, 0.970, 0.780, 0.8]);
;   var famps = Control.names([\gain]).kr([1.0, 0.355, 0.0355, 0.011]);
;
;   var filters = TwoPole.ar(excitation, Lag.kr(ffreqs, sweepRate), Lag.kr(fradii, sweepRate), Lag.kr(famps, sweepRate) );
;
;   Out.ar(out, amp*Mix(filters) );
; }).store;
;
; v = Synth(\voicform, target: s)
; v.set("freq", 100);
; v.free;


; SynthDef.new("mcldjospiano1", { | out = 0, freq = 440, gate = 1,
; amp=0.1, pan = 0|
;         var impresp, imps, dels, hammerstr, velocity, string, ampcomp,
; pldelay, cutoff;
;         velocity = Latch.kr(gate, gate);
;
;         cutoff = EnvGen.kr(Env.asr(0.00001, 1, 0.2, curve: -4), gate,
; doneAction: 2) * 15000 + 30;
;
;         // We start off by appromixating the piano's impulse response.
;         impresp = WhiteNoise.ar(1, 0, EnvGen.ar(Env.perc(0.02, 0.02)));
;         impresp = LPF.ar(impresp, freq.expexp(50, 1000, 10000, 500));
;         // FreeVerb is NOT a piano soundboard impulse response! Just a standin
;         impresp = FreeVerb.ar(impresp, 0.8, freq.linlin(300, 600, 0.1, 0.9),
; freq.linlin(300, 600, 0.19, 0.01));
;         impresp = LeakDC.ar(impresp);
;
;         // Then we simulate the multiple strikes of the hammer
; against the string
;         dels = #[0.002, 0.006, 0.009] * freq.explin(100, 1000, 1, 0.01);
;         imps = DelayN.ar(impresp, dels, dels, #[0.85, 0.32, 0.22]);
;         // Note: at higher velocity, the LPF goes higher, making the
; hammer hits more pointy & separate
;         imps = LPF.ar(imps, freq * 2 * #[1, 1.5, 1.5] * velocity * 2, mul: 8);
;         hammerstr = imps.sum;
;
;         // Now push the sound into Pluck, to simulate the string vibration
;         pldelay = (freq * [Rand(0.999, 0.9995), 1, Rand(1.00005,
; 1.001)]).reciprocal;
;         string = Pluck.ar(hammerstr, Impulse.kr(0.000001), pldelay,
; pldelay, 10.5, 0.4);
;         string = LeakDC.ar(string).sum;
;
;         // patch gives un-piano-like amplitude variation across
; pitch; let's compensate
;         ampcomp = freq.max(350).min(1000).linlin(350, 1000, 1, 60);
;         string = string * ampcomp;
;
;         // filter is to damp the string when the note stops
;         string = LPF.ar(string, cutoff);
;
;         Out.ar(out, Pan2.ar(string, pan, (amp * 10)));
; }).store;
;
;
; SynthDef(\piano, { arg outBus, freq, amp, dur, pan;
;         var sig, in, n = 6, max = 0.04, min = 0.01, delay, pitch, detune, hammer;
;         hammer = Decay2.ar(Impulse.ar(0.001), 0.008, 0.04,
;             LFNoise2.ar([2000,4000].asSpec.map(amp), 0.25));
;         sig = Mix.ar(Array.fill(3, { arg i;
;                 detune = #[-0.04, 0, 0.03].at(i);
;                 delay = (1/(freq + detune).midicps);
;                 CombL.ar(hammer, delay, delay, 50 * amp) +
;                 SinOsc.ar(
;                     [(freq * 2) + SinOsc.kr(2, Rand(0, 1.0), 4),
;                     freq * [4.23, 6.5]].flat ,
;                     0,
;                     amp * [0.1, 0.25, 0.3]).sum
;                 }) );
;
;
;         sig = HPF.ar(sig,50) * EnvGen.ar(Env.perc(0.0001,dur, amp, -1), doneAction:2);
;         Out.ar(outBus, Pan2.ar(sig,pan));
; }).send(s);
